### DISCUSSION TITLE
Instance Segmentation Datasets Overview - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#3376] [Docs]
# Instance Segmentation Datasets Overview - Ultralytics YOLOv8 Docs

Learn about the Ultralytics YOLO dataset format for segmentation models. Use YAML to train Detection Models. Convert COCO to YOLO format using Python.

https://docs.ultralytics.com/datasets/segment/

**Comment by GLawrence96** [comment]
If each line in the segmentation data set file is formatted as n x and y coordinates, why is there not an even number of coordinates on each line? The example given with a label of 0 only has 9 datapoints.

Thanks in advance!

**Reply by pderrenger** [reply]
@GLawrence96 hello! The segmentation datasets in YOLOv8 use a polygonal annotation format where each object instance is represented by a series of points outlining the object's shape. Each point is defined by an x and y coordinate, so typically, you would expect an even number of values representing these points.

However, if you encounter an odd number of coordinates, it could be due to a few reasons:

1. **Annotation Error**: There might be a mistake in the annotation file where a coordinate is missing or an extra number is added. It's important to verify the integrity of the dataset and correct any errors.

2. **Special Cases**: In some instances, additional values might be included for specific purposes, such as a class label or an instance identifier, which could lead to an odd total count. Make sure to follow the dataset's specific format guidelines.

If you're seeing an odd number of coordinates consistently across your dataset, it's worth double-checking the dataset format and the process used to generate these files. For YOLOv8, the expected format for segmentation should indeed have an even number of coordinates following the class label, as each pair represents a point (x, y) on the polygon.

If you need further clarification or assistance with your dataset, please refer to our documentation on segmentation datasets or reach out with more details, and we'll be happy to help you resolve the issue. 😊🛠️

**Comment by sbmalik** [comment]
Kindly correct the **Usage** guide for segmentation from
```
# Start training from a pretrained *.pt model
yolo detect train data=coco128-seg.yaml model=yolov8n-seg.pt epochs=100 imgsz=640
```

to (detect to segement)


```
# Start training from a pretrained *.pt model
yolo segment train data=coco128-seg.yaml model=yolov8n-seg.pt epochs=100 imgsz=640
```

**Reply by pderrenger** [reply]
@sbmalik thank you for pointing out the typo in the documentation. The correct usage guide for training a segmentation model with YOLOv8 should indeed use the `segment` command rather than `detect`. The updated command should look like this:

```python
# Start training from a pretrained *.pt model
yolo segment train data=coco128-seg.yaml model=yolov8n-seg.pt epochs=100 imgsz=640
```

I'll make sure this correction is reflected in the documentation. Your attention to detail is much appreciated! 🙌 If you have any more questions or need further assistance, feel free to ask. Happy coding!

**Comment by ShriNear** [comment]
I have all the annotations in one single file how to convert that to yolo format?

**Reply by pderrenger** [reply]
@ShriNear to convert your single-file annotations to the YOLO format, you'll need to create a separate annotation file for each image, where each annotation file contains the bounding box and class information for the objects in the corresponding image.

Here's a general outline of the steps you might take:

1. **Parse Your Annotation File**: Read your single annotation file and parse the contents to extract the annotations for each image. The specifics of this step will depend on the format of your annotation file.

2. **Convert Annotations to YOLO Format**: For each image's annotations, convert the bounding box coordinates and class labels to the YOLO format. In YOLO, each line in an annotation file represents an object and contains the class index followed by the normalized bounding box coordinates (center_x, center_y, width, height), all separated by spaces.

3. **Write Annotations to Separate Files**: For each image, write the converted annotations to a new text file with the same base name as the image file, but with a `.txt` extension. Place these files in a directory, typically named `labels`.

4. **Create a Dataset YAML File**: Create a YAML file that specifies the paths to your training and validation images and labels, as well as the number of classes and class names.

Here's an example of what the YOLO-formatted annotation might look like for one object in an image:

```
0 0.486 0.830 0.172 0.320
```

This represents an object of class `0` with a bounding box centered at `(0.486, 0.830)` and with a width of `0.172` and a height of `0.320`, all normalized to the dimensions of the image.

And here's an example of a dataset YAML file:

```yaml
train: /path/to/train/images
val: /path/to/val/images

nc: 80
names: ['class1', 'class2', ..., 'class80']
```

Remember to adjust the paths and class names to match your dataset.

For more detailed instructions and examples, please refer to the documentation on preparing your data for training with YOLOv8. The documentation provides comprehensive guidance on formatting your dataset correctly. 📚

**Comment by jvishwa06** [comment]
I'm Facing Error Help me with this .
RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 15 but got size 0 for tensor number 1 in the list.

**Reply by pderrenger** [reply]
@Vishu-06 it seems like you're encountering a `RuntimeError` due to a size mismatch in your tensors, specifically in dimension 1. This type of error often occurs during operations that expect tensors to be of the same size in all dimensions except for the one specified (in this case, dimension 1).

To address this issue, please ensure the following:
- Your batched tensors should have the same size in all dimensions except for the batch dimension (usually dimension 0, but here it's mentioned as dimension 1).
- Check the shapes of the tensors involved in the operation that's causing the error. You can print out the shapes before the operation to debug.
- If you're concatenating tensors, make sure that they are compatible in all dimensions except the one along which you're concatenating.
- Verify that your dataset is correctly loaded and that there are no empty tensors or mismatched dimensions due to incorrect data processing or augmentation.

If you're using a custom dataset, ensure that it's properly formatted according to the YOLOv8 dataset requirements for segmentation tasks. If you've recently made changes to your dataset or data loading code, review those changes for potential issues.

If the problem persists, you might want to provide more context or share a snippet of the code where the error occurs, so that the community can offer more targeted assistance. Remember to check the data loading and preprocessing steps carefully, as they are common places where tensor size mismatches can happen. 🛠️

**Comment by Thiago-MM** [comment]
Supposing I have a discontinuous object, would it be possible to assemble a dataset that accepts two polygons referring to the same object? In this tutorial it is apparently only shown what the format should be like for polygons, but what about for discontinuous objects?

**Reply by pderrenger** [reply]
@Thiago-MM yes, it's possible to assemble a dataset that includes discontinuous objects represented by multiple polygons. In the Ultralytics YOLO format for segmentation, each polygon is associated with an object instance. If an object is discontinuous and consists of multiple parts, you can simply include multiple polygons for that object instance in your dataset.

When annotating your dataset, assign the same class and instance ID to each polygon that belongs to the same discontinuous object. This way, during training, the model will learn to recognize these multiple polygons as parts of a single object.

Here's a simplified example of how you might represent a discontinuous object with two polygons in your dataset:

```yaml
- image_id: '1'
  class_id: '0'
  segmentation: [[x1, y1, x2, y2, ...], [x3, y3, x4, y4, ...]] # Two polygons for the same object
```

In this example, both polygons are part of the same list under the `segmentation` key, indicating that they belong to the same object instance. Make sure to follow the documentation on segmentation datasets for more detailed instructions on preparing your data. 📘🔍

**Comment by karlhm76** [comment]
It is important to realise that the detection model used for auto generating segmentation annotations must be a YOLOv8 detection model. I had YOLOv5 model and it would not work. I had to train a YOLOv8 detection model first.

**Reply by pderrenger** [reply]
@karlhm76 that's correct! For generating segmentation annotations using a detection model, it's essential to use a YOLOv8 detection model. The architecture and features in YOLOv8 are specifically designed to work seamlessly with the segmentation tasks in the Ultralytics ecosystem. If you previously used a YOLOv5 model, you would indeed need to train a YOLOv8 model to proceed with instance segmentation tasks. If you need guidance on training a YOLOv8 model, you can refer to the training documentation for detailed instructions. Keep up the great work, and if you have any more questions or need further assistance, feel free to reach out! 😊🚀
